components/timerProgressBar/index.tsx

```diff
@@ -0,0 +1,72 @@
+import {useEffect, useState} from "react";
+import { Box, Text } from "@chakra-ui/react"
+const { v4: uuidv4 } = require('uuid');
+
+interface ITimerProgressBar {
+    startTime: number;
+    endTime: number;
+    barColor?: string;
+    containerStyles?: object;
+    updateFrequencyMillis?: number;
+};
+
+const defaultContainerStyles = {
+    height: 40,
+    width: 80,
+    backgroundColor: "#c5bfd4",
+    borderRadius: "10%",
+    margin: 50,
+};
+
+/**
+ * A component which displays the progress of the current system time between startTime and endTime
+ * @param barColor: optional background color for the bar
+ */
+export default function TimerProgressBar({startTime, endTime, barColor="#68D391", containerStyles=defaultContainerStyles, updateFrequencyMillis=1000} : ITimerProgressBar) : React.ReactElement {
+
+    const [completed, setCompleted] = useState(0); // % completion
+    const [id, setId] = useState(null); // an ID assigned randomly when the input parameters are changed (to reset the timer)
+
+    const tick = (uuid: string) => {
+        if(uuid != id) return; // cancel if a new timer has been set
+
+        //set the completed value to be the percentage between startTime and endTime
+        const now = new Date().getTime();
+        const timeSinceStart = now - startTime;
+        if(timeSinceStart >= 0 && endTime > 0) {
+
+            //calculate percentage of timeDiff met
+            const timeDiff = endTime - startTime;
+            const percentageDone = timeSinceStart / timeDiff * 100;
+            setCompleted(Math.min(percentageDone, 100));
+            if(percentageDone >= 100) {
+                return; // timer done
+            }
+        }
+
+        //reschedule self
+        setTimeout(() => tick(uuid), updateFrequencyMillis);
+    }
+
+    // if startTime or endTime has changed then re-initialize the bar
+    useEffect(() => {
+        const uuid = uuidv4();
+        setTimeout(() => setId(uuid), 1000);
+    }, [startTime, endTime]);
+
+    useEffect(() => {
+        if(id == null) return;
+        setTimeout(() => tick(id), updateFrequencyMillis);
+    }, [id]);
+
```

Comment:
This was quite hard to get my head around, and I think it might have a few bugs (or at least react's exhaustive-deps plugin would complain about not including `tick` or `updateFrequencyMillis` in the second useEffect call).

It looks like it's job is to dynamically update the progress of the bar, as time goes on? If so, how about
```
  const [lastUpdateTime, setLastUpdateTime] = useState(new Date().getTime());

  useEffect(() => {
    setLastUpdateTime(new Date().getTime());

    const updateLastUpdateTime = setInterval(() => {
      const newLastUpdateTime = new Date().getTime();
      setLastUpdateTime(newLastUpdateTime);

      if (newLastUpdateTime > endTime) {
        clearInterval(updateLastUpdateTime);
      }
    }, updateFrequencyMillis);

    return () => clearInterval(updateLastUpdateTime)
  }, [startTime, endTime, updateFrequencyMillis, setLastUpdateTime]);

  const calculatePercentage = () => {
    const timeSinceStart = lastUpdateTime - startTime;

    if (timeSinceStart === 0) {
      return 0;
    }

    if (endTime === 0) {
      return 100;
    }

    const timeDiff = endTime - startTime;
    return Math.min(timeSinceStart / timeDiff, 1) * 100;
  }

  const completed = calculatePercentage();
```

It's tracking the lastUpdateTime instead of right now so if the component re-renders, we don't render more than `updateFrequencyMillis` tells us to. In this solution useEffect is only responsible for the side effect part outside React (i.e. periodically setting the time to get React to re-render and update the bar), and the rest is left inside the usual render method.

Reply:
Yes exactly, sorry about the confusing code

I think that it's complicated because of the `id` malarky, which I was using to prevent multiple intervals from running at the same time. In your code I think this is fulfilled by this line?

```
return () => clearInterval(updateLastUpdateTime)
```

I don't really understand what is being returned here, it's returning a pointer to a function but when would it call `clearInterval`? Actually I didn't know that the `useEffect` hook returned anything :sweat_smile: also I don't understand why `setLastUpdateTime` is in the useEffect params? A priori this means that it would run if the `setLastUpdateTime` function was changed?

I agree that it's a lot cleaner